### PR TITLE
Make FFI support optional, change APIs to be `safe` (#2302)

### LINKED ISSUE #2301: ffi::ArrowArray::try_new is safe
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

`ffi::ArrowArray::try_new` is marked unsafe, this is despite being called by the `safe` function `Array::to_raw`.

I initially thought this was a mistake in `Array::to_raw` but I think this API shouldn't actually be marked unsafe. In particular **leaking memory is not unsafe**, and this appears to be the only consequence from misuse.

**Describe the solution you'd like**

We should make `ffi::ArrowArray::try_new` safe

**Describe alternatives you've considered**

We could not do this

**Additional context**

### LINKED ISSUE #2302: Make FFI support Optional via Feature Flag
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Many use-cases will not need FFI support, and as such being able to disable FFI support will allow eliminating a large amount of non-trivial, unsafe code from their build

**Describe the solution you'd like**

I would like a feature flag to disable FFI support

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #2301 
Closes #2302

# Rationale for this change
 

Most applications don't need FFI support, and putting this behind a feature flag allows them to reduce their dependency footprint 

# What changes are included in this PR?

# Are there any user-facing changes?

Yes, ffi is no longer enabled by default, and `Array::to_raw` is removed

### CODE REVIEW

File: arrow/src/array/array.rs [hunk 1]
```diff
@@ -733,42 +708,6 @@ fn new_null_sized_decimal(
...
-    let data = src.data();
-    let array = ffi::FFI_ArrowArray::new(data);
-    let schema = ffi::FFI_ArrowSchema::try_from(data.data_type())?;
-
-    std::ptr::write_unaligned(out_array, array);
-    std::ptr::write_unaligned(out_schema, schema);
-
-    Ok(())
-}
-
```
Comment (tustvold) [inline arrow/src/array/array.rs:771]: These are moved to the ffi module to make the feature flags easier

File: arrow/src/buffer/immutable.rs [hunk 2]
```diff
@@ -77,30 +76,6 @@ impl Buffer {
...
-    /// * `data` - An [crate::ffi::FFI_ArrowArray] with the data
-    ///
-    /// # Safety
-    ///
-    /// This function is unsafe as there is no guarantee that the given pointer is valid for `len`
-    /// bytes and that the foreign deallocator frees the region.
-    #[deprecated(
-        note = "use from_custom_allocation instead which makes it clearer that the allocation is in fact owned"
-    )]
-    pub unsafe fn from_unowned(
```
Comment (tustvold) [inline arrow/src/buffer/immutable.rs:96]: This method was already deprecated, so I opted to just remove it

File: arrow/src/ffi.rs [hunk 3]
```diff
@@ -836,10 +838,11 @@ impl<'a> ArrowArrayRef for ArrowArrayChild<'a> {
...
     /// creates a new `ArrowArray`. This is used to export to the C Data Interface.
-    /// # Safety
-    /// See safety of [ArrowArray]
-    #[allow(clippy::too_many_arguments)]
-    pub unsafe fn try_new(data: ArrayData) -> Result<Self> {
+    ///
+    /// # Memory Leaks
+    /// This method releases `buffers`. Consumers of this struct *must* call `release` before
+    /// releasing this struct, or contents in `buffers` leak.
+    pub fn try_new(data: ArrayData) -> Result<Self> {
```
Comment (tustvold) [inline arrow/src/ffi.rs:845]: This method wasn't actually unsafe, and was being called by safe methods previously without any additional checks #2301

File: arrow/src/array/array.rs [hunk 4]
```diff
@@ -216,15 +214,6 @@ pub trait Array: fmt::Debug + Send + Sync + JsonEqual {
         self.data_ref().get_array_memory_size() + std::mem::size_of_val(self)
             - std::mem::size_of::<ArrayData>()
     }
-
-    /// returns two pointers that represent this array in the C Data Interface (FFI)
-    fn to_raw(
```
Comment (tustvold) [inline arrow/src/array/array.rs:221]: I didn't see a compelling reason for this API to exist, ultimately the use of `std::any::Any` means users can't implement `Array` for custom types, and I can't see why arrow-rs would ever make use of this extension point

File: arrow/Cargo.toml [hunk 5]
```diff
@@ -72,14 +72,16 @@ prettyprint = ["comfy-table"]
...
 # target without assuming an environment containing JavaScript.
 test_utils = ["rand"]
-pyarrow = ["pyo3"]
+pyarrow = ["pyo3", "ffi"]
 # force_validate runs full data validation for all arrays that are created
 # this is not enabled by default as it is too computationally expensive
 # but is run as part of our CI checks
 force_validate = []
+# Enable ffi support
+ffi = []
```
Comment (alamb) [inline arrow/Cargo.toml:81]: 🤔  I was going to recommend adding this to the "list of features of this crate" but then it turns out we don't seem to have any docs for them 🤦  

https://docs.rs/arrow/19.0.0/arrow/

File: arrow/src/array/array.rs (hunk 4 above)
Comment (alamb) [reply arrow/src/array/array.rs:221]: I do not feel comfortable reviewing this change -- I think @viirya  / @kszucs  may know more about its usecase

**Review by alamb — COMMENTED** [review]
The idea of an optional flag looks good to me. I don't feel comfortable reviewing the FFI changes as I am not familiar with that code

**Review by viirya — COMMENTED** [review]
Looks like a good direction to me. I will take a close look later.

File: arrow/src/array/array.rs (hunk 4 above)
Comment (viirya) [reply arrow/src/array/array.rs:221]: I think for this `to_raw` usage, `ArrowArray` provides `into_raw` which should be enough.

**Review by viirya — APPROVED** [review]

File: arrow/src/array/array.rs (hunk 4 above)
Comment (viirya) [reply arrow/src/array/array.rs:221]: We only use `ArrowArray.into_raw` actually.

**Pull request merged**